### PR TITLE
Handle partitions natively in W&B IO Manager

### PR DESCRIPTION
**Pull request by chrishiste**
## Summary & Motivation

This PR aims to enrich the W&B integration by enhancing the functionality and user experience of the IO manager. We've focused on native handling of (multi-)partitioned assets, which streamlines the process and ensures our implementation is more robust. We also addressed and resolved several bugs, to provide a more seamless experience to users.

In our commitment to clarity and ease of use, we've added comprehensive examples showcasing the use of partitions and multi-partitions, as well as how to specify a version or alias in the configuration metadata.

We’ve also given much-needed attention to user-facing errors, rewording them for better understanding and visibility. Moreover, we introduced additional errors to alert users when there might be issues with their provided configuration, improving their ability to self-diagnose and rectify issues.

## How I Tested These Changes

The examples added were run in multiple scenarios to ensure that they cover common use cases and successfully demonstrate the intended functionality.

We also replicated common bugs that were reported, and confirmed that the fixes eliminated these issues.

Lastly, we ensured backward compatibility and non-disruptive deployment with integration tests.

We're excited to share these updates and we believe these improvements will significantly enhance the user's experience. Looking forward to your feedback, Dagster team!

### CODE REVIEW

**Comment by jamiedemaria** [comment]
I think @yuhan will be best situated to review this PR, but she is out on vacation until mid next week!

File: examples/with_wandb/with_wandb/assets/advanced_partitions_example.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+                            "get": "default_table_name",
+                        },
+                    },
+                },
+            },
+        )
+    },
+    output_required=False,
+)
+def read_objects_directly(context, partitions):
```
Comment (yuhan) [inline examples/with_wandb/with_wandb/assets/advanced_partitions_example.py:74]: trying to wrap my head around here. where does the `partitions` come from?

**Comment by yuhan** [comment]
the test is mostly green except for the formatting/linting. could you run `make ruff` to lint the code to be consistent with the rest of codebase?

File: examples/with_wandb/with_wandb/assets/advanced_partitions_example.py (hunk 1 above)
Comment (chrishiste) [reply examples/with_wandb/with_wandb/assets/advanced_partitions_example.py:74]: From the `AssetIn`

**Comment by chrishiste** [comment]
@yuhan Code is linted, I had to pull the latest master to get the new config :)

**Review by yuhan — APPROVED** [review]

**Comment by yuhan** [comment]
The build failure should have been resolved on master. I tested this PR + latest master here: https://github.com/dagster-io/dagster/pull/15758 the build is green.

**Pull request merged**